File: app/src/main/java/com/upsage/evosummermobilelab/data/singleton/NotesData.java
package com.upsage.evosummermobilelab.data.singleton;

import android.content.Context;

import androidx.paging.DataSource;
import androidx.room.Room;

import com.upsage.evosummermobilelab.data.entries.Note;
import com.upsage.evosummermobilelab.data.intefaces.AppDatabase;

import java.util.Date;


public class NotesData {
    private static NotesData instance;
    private final AppDatabase db;

    private NotesData(Context context) {
        db = Room.databaseBuilder(context.getApplicationContext(),
                AppDatabase.class, "database")
                .allowMainThreadQueries()
                .build();
    }

    public static void initialize(Context context) {
        if (instance == null) {
            synchronized (NotesData.class) {
                if (instance == null)
                    instance = new NotesData(context);
            }
        }
    }

    public static NotesData getInstance() {
        return instance;
    }

    public Note getItem(int position) {
        return db.userDao().loadById(position);
    }

//    private void addNote(Note note) {
//        db.userDao().insertAll(note);
//    }

    public Note getNewNote() {
        Note note = new Note(new Date(), "");
        //addNote(note); should not add empty note to db
        return note;
    }

    public void delete(int noteId) {
        db.userDao().delete(noteId);
    }

    public void save(Note note) {
        if (note.getDescription().isEmpty()) {
            db.userDao().delete(note.getId());
        } else {
            note.setUpdateDate(new Date());
            if (db.userDao().insertAll(note)[0] == -1)
                db.userDao().update(note);
        }
    }

    public DataSource.Factory<Integer, Note> getLike(String str) {
        return db.userDao().findByDescription("%" + str + "%");
    }

    public DataSource.Factory<Integer, Note> getAllPagedAsc() {
        return db.userDao().getAllPagedAsc();
    }

    public DataSource.Factory<Integer, Note> getAllPagedDesc() {
        return db.userDao().getAllPagedDesc();
    }
}

File: app/src/main/java/com/upsage/evosummermobilelab/data/NotesData.kt
package com.upsage.evosummermobilelab.data

import android.content.Context
import androidx.paging.DataSource
import androidx.room.Room
import com.upsage.evosummermobilelab.data.intefaces.AppDatabase
import java.util.*


class NotesData private constructor(context: Context) {
    private val db: AppDatabase = Room.databaseBuilder(context.applicationContext,
            AppDatabase::class.java, "database")
            .build()


    val newNote: Note
        get() = Note(Date(), "")

    val allPagedAsc: DataSource.Factory<Int, Note>
        get() = db.userDao().allPagedAsc

    val allPagedDesc: DataSource.Factory<Int, Note>
        get() = db.userDao().allPagedDesc

    fun delete(note: Note) {
        db.userDao().delete(note)
    }

    suspend fun save(note: Note) {
        if (note.description.isEmpty()) {
            db.userDao().delete(note)
        } else {
            note.updateDate = Date()
            if (db.userDao().insertAll(note)[0] == -1L)
                db.userDao().update(note)
        }
    }

    fun getLike(str: String): DataSource.Factory<Int, Note> {
        return db.userDao().findByDescription("%$str%")
    }

    companion object {
        var instance: NotesData? = null
            private set

        fun initialize(context: Context) {
            if (instance == null) {
                synchronized(NotesData::class.java) {
                    if (instance == null)
                        instance = NotesData(context)
                }
            }
        }
    }
}
